### PR TITLE
feat(mongodb4): added mongodb4 instrumentation

### PR DESCRIPTION
**Pull request by osherv**
MongoDB4 Instrumentation!

### CODE REVIEW

**Comment by dyladan** [comment]
Is there a reason this is a new instrumentation instead of updating the existing mongodb instrumentation?

**Comment by osherv** [comment]
@dyladan 
Yes :) Here's the related Issue:
https://github.com/open-telemetry/opentelemetry-js-contrib/issues/577
'I think we should support both of them since v3 is still massively used'

**Comment by dyladan** [comment]
I was not suggesting to drop support for v3, but to add support for v4 to the existing instrumentation using a new `InstrumentationNodeModuleDefinition`.

**Comment by osherv** [comment]
@dyladan They were pretty much changed everything from 3 to 4. That will cause collisions in the depenencies.
I think it has to be in another package. Or, does otel have support for multiple versions dependencies in the same package?

**Comment by osherv** [comment]
@tydhot 
I tried, but the new instrumentation won't be applyed
Screen Shot 2021-12-22 at 17 42 54
lied in that case.
Screen Shot 2021-12-22 at 17 44 09

Is there any option to fix that without create a new package?

**Comment by dyladan** [comment]
In most other instrumentations where we patch multiple major versions, we only use the types from the latest version and use `any` or `unknown` or hand-rolled types for the other versions.

**Comment by osherv** [comment]
@dyladan Thanks! ill try!

**Comment by osherv** [comment]
@dyladan Hey :)
The issue is almost ready!
Two last fails:
1. mongo4 is not supported for node < 12
2. Code coverage is now lower because the mongo3 tests is only under test-all-versions and not under tests.

Any idea how to handle with these? Thanks alot

**Review by Rauno56 — APPROVED** [review]
Looks good! Thanks!

Please remove `console.log(getTestSpans());` from `mongodb-v3.test.ts` as well.

Some of the small changes(few newlines, typecast to any) could also be taken to v3 test file, but that's not a deal breaker.

**Comment by Rauno56** [comment]
`mongodb@4` doesn't support node 8 and 10. Please ignore the package in the CI matrix via `lerna-extra-args` in `unit-test.yml`.

**Comment by osherv** [comment]
@Rauno56 Fixed everything! Thanks a lot for the review :)

**Comment by Rauno56** [comment]
Nice! Even the smallest of my nits now addressed! :+1: I'd love to get another review before we merge.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts
Comment (dyladan) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts:168]: I think this should just be an interface or a type. Its concrete implementation is never used.

Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts:168]: Agree! fixed

**Review by dyladan — APPROVED** [review]

**Comment by Rauno56** [comment]
:exclamation: This cannot yet be merged. For some reason, the tests stay pending forever. Work locally.

**Review by dyladan — CHANGES_REQUESTED** [review]
@Rauno56 please use "request changes" to block a PR so that it is not accidentally merged

File: plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts:19]: You cannot import from your instrumented library (mongodb)
This will crash applications that installed the instrumentation but not `mongodb` package (valid use case).
Use `import type * as mongodb from 'mongodb';`
This will remove the import fromo JS after transpilation.

File: .nycrc
Comment (blumamir) [inline .nycrc:12]: why ignore the test coverage on test files?

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/package.json:63]: We are installing mongodb v4 as dev dependency, which means that v3 tests will no pass, only in tav, right?

**Review by blumamir — CHANGES_REQUESTED** [review]
Wrote a few comments.
The only critical thing is not to `import * from 'mongodb'` as it can potentially break users. 
Other comments are recommended but not a blocker

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/package.json:10]: this change means we are running PullRequests CI tests only for v4, and stop running v3 tests on PRs.
https://github.com/open-telemetry/opentelemetry-js-contrib/blob/81b3190af64bda14f87c5b0cbd6172bafda26408/package.json#L28
something we might want to consider...

Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:10]: That is correct. I can't think about any other option. Suggestions?

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:63]: Right :)

File: .nycrc
Comment (osherv) [reply .nycrc:12]: They are not representing code coverage. That caused my to fail duo to low code coverage.

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:63]: What do you think about documenting it somehow?
Maybe we can write a comment at the top of the v3 tests file with instructions on how to run them locally.
The next person who will try to run v3 test will see all tests failing and be surprised

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:503]: Fixed ty!

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:499]: Fixed ty!

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:479]: Fixed ty!

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:10]: Not a particularly good one
We can run `tav` here as well, possibly with just 2 versions. But that sounds clumsy and complex.
Let's wait for more minds on this one @Rauno56 @dyladan

Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:10]: btw, The tests are only can run locally (we need mongodb server). So anyway i dont think i would change much

File: plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/types.ts:19]: Good one! fixed

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:503]: amazing.
If you are up to it, you can also further shorten it to:
```js
const commandObj = command?.documents?.[0] ?? command?.cursors ?? command
```

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:584]: another small nit: empty strings are falsy too, so you can just write `if (host && port)`

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:557]: same here. a shortened writing of this code could look like this:
```js
const commandObj = command?.query ?? command?.q ?? command
```

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:503]: I would maybe think about putting each of the three possibilities on its own line for readability

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:10]: Yes, I usually spin up a mongo server locally with docker and run the tests against it.
CI tests also start a mongo server [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/81b3190af64bda14f87c5b0cbd6172bafda26408/.github/workflows/unit-test.yml#L27), so the tests can pass.

File: .nycrc
Comment (blumamir) [reply .nycrc:12]: I understand the problem but believe we should find a proper solution to this than disable test coverage altogether on this package.
It's related to the issue of not running v3 tests in `npm run test`.
Maybe when it is resolved, this one will be too,

Comment (blumamir) [reply .nycrc:12]: never mind my comment, I read it again and understood why it works.
sorry for the trouble :)

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:503]: Done

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:292]: Dictionary like object with the command details

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:584]: Fixed

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:557]: Fixed

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:63]: Added doc :)

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:584]: Fixed

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:498]: `command?.cursor`

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:497]: what is the meaning of this documents in this place? can it be empty array?

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:292]: Can you please check how this object is constructed and if keys[0] is guraenteed to contain the command type?

**Comment by Rauno56** [comment]
MongoDB tests fail, but do not terminate the process. That's why the tests never finish:
https://github.com/Rauno56/opentelemetry-js-contrib/runs/4760394725?check_suite_focus=true

**Comment by Rauno56** [comment]
Regarding :point_up: That definitely needs to be fixed in two aspects:

1. errors from the tests should always terminate the process. Some combination of MongoDB and mocha doesn't currently work well, because the test is marked failed by mocha internally but it doesn't really throw and terminate. I tried it with the instrumentation disabled(but the rest of the setup there) - same thing. To reproduce just throw in the callback of the mongo command. The only thing that comes to mind right now is to catch errors in callbacks and call `done(err);` ourselves. :disappointed: 
2. The thrown error comes from the fact that mongo resolves `localhost` and stores the IP on the field we check against `localhost`. To bypass that, we should set the host to `127.0.0.1` from the get-go to avoid the discrepancy. That can be done

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:10]: We talked about this in the SIG. We recommend to test the most popular major version (in this case v3) and test other versions with TAV.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:498]: Fixed

**Comment by zombispormedio** [comment]
any update about this PR? I am interested in

**Comment by osherv** [comment]
> any update about this PR? I am interested in

Sry for the delay, i had to omicron virus -_-..
I'm working on it right now

**Comment by zombispormedio** [comment]
> Sry for the delay, i had to omicron virus -_-.. I'm working on it right now

Sorry, just wondering. I hope you had a good recovery. Thank you.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:497]: Fixed! ty

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:292]: Hey, i looked all over their code and didnt saw any case that this is not the command type.
but, this is Document so nothing guraenteed. Do you want me to add verification?

**Comment by osherv** [comment]
@Rauno56  @blumamir 
Hey! Good to be back after the Omicron XD.
Fix all of you comments. I need to fix my CLA, so the moment you approves my PR, ill open new branch with this changes.
:)

**Comment by blumamir** [comment]
> @Rauno56 @blumamir Hey! Good to be back after the Omicron XD. Fix all of your comments. I need to fix my CLA, so the moment you approves my PR, ill open new branch with these changes. :)

Good to have you back and thanks for the great work. Hope you are feeling well now :)
You can also open a new PR with the fixed CLA and I'll finish the review there. It will probably take me a day or two to get to it.

File: .github/workflows/unit-test.yml
Comment (blumamir) [inline .github/workflows/unit-test.yml:24]: since we now run tests for version 3 in CI, we can keep running it on both node 8 and 10

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:491]: remove console.log

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:288]: I think you are trying to filter out [this method](https://docs.mongodb.com/v4.2/reference/method/db.isMaster/)
According to the docs, it is deprecated and the new method is [hello](https://docs.mongodb.com/v4.2/reference/method/db.hello/#db.hello).

[This is the relevant part from the driver](https://github.com/mongodb/node-mongodb-native/blob/dd81438593f9900122ed16a583eb9c84d0408b76/src/cmap/connect.ts#L231):
```js
    [serverApi?.version ? 'hello' : LEGACY_HELLO_COMMAND]: true,
```

[where](https://github.com/mongodb/node-mongodb-native/blob/dd81438593f9900122ed16a583eb9c84d0408b76/src/constants.ts#L122)
```js
export const LEGACY_HELLO_COMMAND = 'ismaster';
```

I suggest adding `hello` as well to the check

File: .github/workflows/unit-test.yml
Comment (osherv) [reply .github/workflows/unit-test.yml:24]: Fixed :)

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:491]: Oh -_- Fixed

File: plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts:288]: Good point, Fixed

**Comment by blumamir** [comment]
Thanks for addressing all the comments and for adding the support for v4 🥇 
After the CLA is fixed, this can be merged as far as I am concerned

**Review by blumamir — APPROVED** [review]

**Comment by osherv** [comment]
Thanks! next pr will be muck more quicker :)
I'll open new pr with the fixed CLA

**Comment by blumamir** [comment]
closing as this PR migrated to #869 due to CLA issues

**Pull request closed without merging**
